sparkmagic/proposal.md

```diff
@@ -0,0 +1,49 @@
+# Proposal for Incubation
+
+## Subproject name
+
+sparkmagic
+
+## Development team and Advocate
+
+Subproject development team:
+
+* Alejandro Guerrero Gonzalez, Microsoft (`@aggFTW`)
+* Joel Zambrano, Microsoft (`@djoelz`)
+* Auberon Lopez, Cal Poly (`@alope107`)
+
+Steering Council Advocate
+
+* Brian Granger (`@ellisonbg`)
+
+## Subproject goals, scope and functionality
+
+Goals:
+* Make Spark usage from the IPython notebook as frictionless as possible by decoupling Spark code execution from Spark installation.
+* These magics will be packaged so that they can be pip-installable or conda-installable.
+* Allow any Jupyter installation to point to different Spark installations.
+
+Scope:
+* Spark will be the first back end provided by these magics, but it could easily be extended to other big data back ends.
+* Remote Spark execution through a REST endpoint which allows for Scala, Python, and R support as of Sept 2015.
+
+Functionality:
+* Enable users to remotely connect to a REST Spark back end.
+* Return of pandas dataframes from Spark when appropriate. This will allow users to integrate spark with the Python libraries available for pandas dataframes.
+
+## Audience
+
+Spark users who want to have the best experience possible from a Notebook. These users care about the libraries ecosystem the Jupyter community has created.
+
+## Other options
+
+Alternatives we know of:
+
+* Combination of IPython, R kernel (or rpy2), and Scala kernel for an in-Spark-cluster Jupyter installation. This does not allow the user to point to different Spark clusters. It might also result in resource contention (CPU or memory) between the Jupyter installation and Spark.
+* IBM's Spark kernel does not provide a REST endpoint, requires the installation of Jupyter in the cluster, and does not create pandas dataframes.
```

Comment:
Awesome discussion! Thanks so much for your contributions. 

Let me first try to address how Spark works to get to a shared understanding. If I've misunderstood something, please let me know.

The spark driver is the program that holds the spark context and creates the metadata around the  RDDs, which are then evaluated by worker nodes. This driver DOES need network access to nodes in the cluster.

So, as long as you have an executable that has network access to the master in the cluster, for the drivers it creates, and can ask the master to do work, you'll be able to use Spark. This executable could be spark-submit, the pyspark shell, Livy, an IPython kernel, a Scala kernel, IBM's Spark kernel...

The architecture we are proposing for these magics is one such that the magics would talk to a REST endpoint (we are thinking Livy), that can create different drivers for the user and could fetch results for the user. You send a string of [python, scala, R] code that Livy relays to the Spark driver it created and start getting the results back from Livy via over http. You could certainly choose to have a remote kernel that creates your drivers for you and use 0mq, but it's our believe that using a REST endpoint would extend better for other apps other than Jupyter kernels.

The benefits of having a remote installation of Jupyter that is able to connect to different clusters by virtue of changing the URL endpoint is manyfold:

1. Your notebooks are valuable by themselves, without the need to have a cluster running. You can look at them and play with them if the cluster is gone.
2. Any number of users could point to the same endpoint from their notebooks. They might have custom installs of Jupyter or other kernels running for them.
3. Resource contention between Jupyter's kernels and Spark is gone. There might be resource contention between the driver and the master too but that's a different topic, and @freeman-lab 's solution definitely works.

There are challenges with the remote spark submission scenario, like figuring out the right amount of data from the result set to bring back to the client via the wire (regardless of the protocol used). Is it a sample or the top of the result set? I believe we'll have to work through these challenges regardless of the implementation chosen for the remote submission scenario.

Now, the beauty of doing something like this is that, under the hood, this is all http requests via Python. The user can be typing Scala or R code that gets executed in the cluster, but locally, it's all Python code and there's no need to improve multiple kernels to support integration with different languages for Spark. It is the purpose of these magics to allow users to do automatic rich visualizations of their computations in Spark (think Zeppelin, which is written in Scala but supports different languages) by integrating with the library that @ellisonbg is writing the proposal for. Advanced users might want to interact with the raw pandas dataframes that the kernel is using to visualize the data, but that's up to them. Canonical Spark users would do their computations in the cluster (keeping data and compute close together) and only retrieve the end result of their computations.

Again, loving the conversation, and I'm very eager to hear your feedback!